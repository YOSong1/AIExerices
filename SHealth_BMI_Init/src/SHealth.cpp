#include <cstdlib>
#include <cstring>
#include <fstream>
#include <iostream>
#include <vector>
#include <sstream>
#include "SHealth.h"

namespace {
    // 문자열 트리밍 함수
    std::string Trim(const std::string& str) {
        size_t first = str.find_first_not_of(" \t\r\n");
        if (first == std::string::npos) return "";
        size_t last = str.find_last_not_of(" \t\r\n");
        return str.substr(first, (last - first + 1));
    }
    
    // Split 함수를 내부 함수로 이동 (트리밍 추가)
    std::vector<std::string> Split(const std::string& line, char delimiter) 
    {
        std::istringstream iss(line);  
        std::string token;            
        std::vector<std::string> tokens;
        while (std::getline(iss, token, delimiter)) {
            std::string trimmed = Trim(token);
            tokens.push_back(trimmed);
        }
        return tokens;
    }
    
    // 안전한 정수 파싱
    int SafeParseInt(const std::string& str, int defaultValue = 0) {
        try {
            if (str.empty()) return defaultValue;
            return std::stoi(str);
        } catch (const std::exception&) {
            return defaultValue;
        }
    }
    
    // 안전한 실수 파싱
    double SafeParseDouble(const std::string& str, double defaultValue = 0.0) {
        try {
            if (str.empty()) return defaultValue;
            return std::stod(str);
        } catch (const std::exception&) {
            return defaultValue;
        }
    }
    
    // 데이터 유효성 검증
    bool IsValidAge(int age) {
        return age > 0 && age <= 150;
    }
    
    bool IsValidWeight(double weight) {
        return weight > 0.0 && weight <= 1000.0; // 현실적인 범위
    }
    
    bool IsValidHeight(double height) {
        return height > 0.0 && height <= 300.0; // 현실적인 범위 (cm)
    }
    
    // BMI 분류 상수
    constexpr double BMI_UNDERWEIGHT_THRESHOLD = 18.5;
    constexpr double BMI_NORMAL_UPPER = 23.0;
    constexpr double BMI_OVERWEIGHT_UPPER = 25.0;
    
    // 레거시 타입 코드 상수 (기존 호환성 유지)
    constexpr int TYPE_UNDERWEIGHT = static_cast<int>(BmiType::Underweight);
    constexpr int TYPE_NORMAL = static_cast<int>(BmiType::Normal);
    constexpr int TYPE_OVERWEIGHT = static_cast<int>(BmiType::Overweight);
    constexpr int TYPE_OBESITY = static_cast<int>(BmiType::Obesity);
}



int SHealth::CalculateBmi(std::string filename)
{
    records.clear();
    std::ifstream file(filename);
    if (!file.is_open()) {
        return 0; // 파일 열기 실패
    }
    
    std::string line;
    // 헤더 라인 건너뛰기
    if (std::getline(file, line)) {
        // 첫 번째 라인은 헤더로 가정
    }
    
    while (std::getline(file, line)) {
        std::string trimmed_line = Trim(line);
        if (trimmed_line.empty()) continue; // 빈 라인 스킵
        
        // 주석 라인 스킵 (# 또는 // 로 시작)
        if (trimmed_line[0] == '#' || trimmed_line.substr(0, 2) == "//") continue;
        
        std::vector<std::string> tokens = Split(trimmed_line, ',');
        if (tokens.size() < 4) continue; // 토큰 개수 검증
        
        // 안전한 파싱과 유효성 검증
        int age = SafeParseInt(tokens[1]);
        double weight = SafeParseDouble(tokens[2]);
        double height = SafeParseDouble(tokens[3]);
        
        // 데이터 유효성 검증 (유효하지 않은 데이터는 건너뛰기)
        if (!IsValidAge(age) || !IsValidHeight(height)) continue;
        
        // 체중이 0이거나 유효하지 않은 경우에도 일단 저장 (나중에 보간)
        double validWeight = IsValidWeight(weight) ? weight : 0.0;
        records.emplace_back(age, validWeight, height);
    }

    // 모듈화된 함수들로 처리 단계 분리
    ImputeMissingWeights();
    CalculateBmiValues();
    CalculateAgeGroupRatios();
    return static_cast<int>(records.size());
}

double SHealth::GetBmiRatio(int age_class, int type) const
{
    auto it = ageGroupRatios.find(age_class);
    if (it == ageGroupRatios.end()) {
        return 0.0; // 해당 연령대 데이터가 없음
    }
    
    const BmiRatios& ratios = it->second;
    
    switch (type) {
        case TYPE_UNDERWEIGHT: return ratios.underweight;
        case TYPE_NORMAL:      return ratios.normal;
        case TYPE_OVERWEIGHT:  return ratios.overweight;
        case TYPE_OBESITY:     return ratios.obesity;
        default:               return 0.0;
    }
}

// 개선된 인터페이스 (BmiType enum 사용)
double SHealth::GetBmiRatio(int age_class, BmiType type) const
{
    return GetBmiRatio(age_class, static_cast<int>(type));
}

int SHealth::GetAgeBucket(int age) const
{
    // 연령대를 10년 단위로 구분
    if (age < MIN_AGE_GROUP) return MIN_AGE_GROUP;
    if (age >= MAX_AGE_GROUP + AGE_GROUP_SPAN) return MAX_AGE_GROUP;
    return (age / AGE_GROUP_SPAN) * AGE_GROUP_SPAN;
}

void SHealth::ImputeMissingWeights()
{
    // 데이터 수집 중 누락된 체중에 나이대(ex. 20대, 30대, 40대 등)의 평균 체중을 적용
    for (int ageGroup : AGE_GROUPS)
    {
        double sum = 0;
        int age_count = 0;
        
        // 해당 연령대의 유효한 체중 데이터 수집
        for (const auto& record : records) 
        {
            if (record.age >= ageGroup && record.age < ageGroup + AGE_GROUP_SPAN)
            {
                if (record.weight > 0.0) {
                    sum += record.weight;
                    age_count++;
                }
            }
        }
        
        // 분모 0 방지 및 평균 체중으로 누락 데이터 보간
        if (age_count > 0) {
            double average_weight = sum / age_count;
            for (auto& record : records) 
            {
                if (record.age >= ageGroup && record.age < ageGroup + AGE_GROUP_SPAN)
                {
                    if (record.weight == 0.0) {
                        record.weight = average_weight;
                    }
                }
            }
        }
    }
}

void SHealth::CalculateBmiValues()
{
    // BMI 계산하기
    for (auto& record : records)
    {
        record.bmi = record.weight / ((record.height / 100.0) * (record.height / 100.0));
    }
}

void SHealth::CalculateAgeGroupRatios()
{
    // 나이대(ex. 20대, 30대, 40대 등)의 BMI기준 저체중, 정상체중, 과체중, 비만 비율 계산
    ageGroupRatios.clear();
    
    for (int ageGroup : AGE_GROUPS)
    {
        int underweight = 0;
        int normalweight = 0;
        int overweight = 0;
        int obesity = 0;
        int sum = 0;
        
        for (const auto& record : records)
        {
            if (record.age >= ageGroup && record.age < ageGroup + AGE_GROUP_SPAN)
            {
                sum++;
                // 기존 로직 유지 (25.0은 어떤 카테고리에도 포함되지 않음)
                if (record.bmi <= BMI_UNDERWEIGHT_THRESHOLD) underweight++;
                else if (record.bmi < BMI_NORMAL_UPPER) normalweight++;
                else if (record.bmi < BMI_OVERWEIGHT_UPPER) overweight++;
                else if (record.bmi > BMI_OVERWEIGHT_UPPER) obesity++;
            }
        }
        
        // 비율 계산 및 저장
        BmiRatios ratios;
        if (sum > 0) {
            ratios.underweight = (double)underweight * 100 / sum;
            ratios.normal = (double)normalweight * 100 / sum;
            ratios.overweight = (double)overweight * 100 / sum;
            ratios.obesity = (double)obesity * 100 / sum;
        }
        // sum이 0인 경우 ratios는 기본값(0.0)으로 초기화됨
        
        ageGroupRatios[ageGroup] = ratios;
    }
}
