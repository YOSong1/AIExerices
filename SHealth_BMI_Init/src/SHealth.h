#ifndef SHEALTH_H
#define SHEALTH_H

#include <string>
#include <vector>
#include <map>
#include <array>

// BMI 분류 타입
enum class BmiType {
    Underweight = 100,
    Normal = 200,
    Overweight = 300,
    Obesity = 400
};

// 연령대 상수
constexpr std::array<int, 6> AGE_GROUPS = {20, 30, 40, 50, 60, 70};
constexpr int MIN_AGE_GROUP = 20;
constexpr int MAX_AGE_GROUP = 70;
constexpr int AGE_GROUP_SPAN = 10;

struct BmiRatios {
    double underweight;
    double normal;
    double overweight;
    double obesity;
    
    BmiRatios() : underweight(0.0), normal(0.0), overweight(0.0), obesity(0.0) {}
};

struct HealthRecord {
    int age;
    double weight;
    double height;
    double bmi;
    
    HealthRecord() : age(0), weight(0.0), height(0.0), bmi(0.0) {}
    HealthRecord(int a, double w, double h) : age(a), weight(w), height(h), bmi(0.0) {}
};

class SHealth {
private:
    std::vector<HealthRecord> records;
    std::map<int, BmiRatios> ageGroupRatios;

    // Helper functions
    void ImputeMissingWeights();
    void CalculateBmiValues();
    void CalculateAgeGroupRatios();
    int GetAgeBucket(int age) const;

public:
    int CalculateBmi(std::string filename);
    double GetBmiRatio(int age_class, int type) const;  // 기존 호환성 유지
    double GetBmiRatio(int age_class, BmiType type) const;  // 개선된 인터페이스
};

#endif
