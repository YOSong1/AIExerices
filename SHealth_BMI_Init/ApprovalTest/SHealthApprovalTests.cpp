#define APPROVALS_GOOGLETEST
#include "ApprovalTests.hpp"
#include <gtest/gtest.h>
#include "SHealth.h"

#include <fstream>
#include <sstream>
#include <iomanip>
#include <cstdio>

using namespace ApprovalTests;

TEST(SHealthApprovalTests, CalculateBmiApproval)
{
    auto approvalsSubdir = Approvals::useApprovalsSubdirectory("approvals");

    std::string input =
        "id,age,weight,height\n"
        "1,25,70,175\n"
        "2,25,48,165\n"
        "3,25,0,165\n"
        "4,27,75,170\n";

    const char* tmp = "temp_input.csv";
    {
        std::ofstream ofs(tmp);
        ofs << input;
    }

    SHealth shealth;
    shealth.CalculateBmi(tmp);

    std::ostringstream oss;
    oss.setf(std::ios::fixed, std::ios::floatfield);
    oss << std::setprecision(1);

    oss << "BMI Ratios for 20s (%)\n";
    oss << "Underweight: " << shealth.GetBmiRatio(20, 100) << "\n";
    oss << "Normal: "      << shealth.GetBmiRatio(20, 200) << "\n";
    oss << "Overweight: "  << shealth.GetBmiRatio(20, 300) << "\n";
    oss << "Obesity: "     << shealth.GetBmiRatio(20, 400) << "\n";

    Approvals::verify(oss.str());

    std::remove(tmp);
}

TEST(SHealthApprovalTests, MultiAgeDistributionApproval)
{
    auto approvalsSubdir = Approvals::useApprovalsSubdirectory("approvals");

    // 20s: 4명 (각 1명씩) -> 25%씩
    // 30s: 4명 (각 1명씩) -> 25%씩
    std::string input =
        "id,age,weight,height\n"
        // 20s (height 170cm)
        "1,25,52.0,170\n"  // BMI ≈ 18.0 -> Underweight
        "2,25,60.7,170\n"  // BMI ≈ 21.0 -> Normal
        "3,25,69.4,170\n"  // BMI ≈ 24.0 -> Overweight
        "4,25,75.1,170\n"  // BMI ≈ 26.0 -> Obesity
        // 30s (height 180cm)
        "5,35,58.3,180\n"  // BMI ≈ 18.0 -> Underweight
        "6,35,68.0,180\n"  // BMI ≈ 21.0 -> Normal
        "7,35,77.8,180\n"  // BMI ≈ 24.0 -> Overweight
        "8,35,84.2,180\n"; // BMI ≈ 26.0 -> Obesity

    const char* tmp = "temp_multi_age.csv";
    {
        std::ofstream ofs(tmp);
        ofs << input;
    }

    SHealth shealth;
    shealth.CalculateBmi(tmp);

    std::ostringstream oss;
    oss.setf(std::ios::fixed, std::ios::floatfield);
    oss << std::setprecision(1);

    oss << "BMI Ratios for 20s (%)\n";
    oss << "Underweight: " << shealth.GetBmiRatio(20, 100) << "\n";
    oss << "Normal: "      << shealth.GetBmiRatio(20, 200) << "\n";
    oss << "Overweight: "  << shealth.GetBmiRatio(20, 300) << "\n";
    oss << "Obesity: "     << shealth.GetBmiRatio(20, 400) << "\n";

    oss << "BMI Ratios for 30s (%)\n";
    oss << "Underweight: " << shealth.GetBmiRatio(30, 100) << "\n";
    oss << "Normal: "      << shealth.GetBmiRatio(30, 200) << "\n";
    oss << "Overweight: "  << shealth.GetBmiRatio(30, 300) << "\n";
    oss << "Obesity: "     << shealth.GetBmiRatio(30, 400) << "\n";

    Approvals::verify(oss.str());
    std::remove(tmp);
}

TEST(SHealthApprovalTests, BoundaryClassificationApproval)
{
    auto approvalsSubdir = Approvals::useApprovalsSubdirectory("approvals");

    // 30s: BMI 경계값 검증 (height 100cm -> BMI == weight)
    // 값: 18.5(저체중 한계), 23.0(과체중 하한 포함), 25.0(어느 범주에도 포함되지 않음)
    std::string input =
        "id,age,weight,height\n"
        "1,35,18.5,100\n"  // Underweight (<=18.5)
        "2,35,23.0,100\n"  // Overweight (>=23 && <25)
        "3,35,25.0,100\n"; // Not counted (obesity is >25 only)

    const char* tmp = "temp_boundary.csv";
    {
        std::ofstream ofs(tmp);
        ofs << input;
    }

    SHealth shealth;
    shealth.CalculateBmi(tmp);

    std::ostringstream oss;
    oss.setf(std::ios::fixed, std::ios::floatfield);
    oss << std::setprecision(1);

    oss << "BMI Ratios for 30s (%)\n";
    oss << "Underweight: " << shealth.GetBmiRatio(30, 100) << "\n";
    oss << "Normal: "      << shealth.GetBmiRatio(30, 200) << "\n";
    oss << "Overweight: "  << shealth.GetBmiRatio(30, 300) << "\n";
    oss << "Obesity: "     << shealth.GetBmiRatio(30, 400) << "\n";

    Approvals::verify(oss.str());
    std::remove(tmp);
}

TEST(SHealthApprovalTests, MissingWeightImputationApproval)
{
    auto approvalsSubdir = Approvals::useApprovalsSubdirectory("approvals");

    // 20s: 체중 0이 평균(비영 0 제외)으로 대체되는지 확인
    // 데이터: (0, 80) -> 평균 80으로 대체되어 모두 비만
    std::string input =
        "id,age,weight,height\n"
        "1,25,0.0,170\n"
        "2,25,80.0,170\n";

    const char* tmp = "temp_imputation.csv";
    {
        std::ofstream ofs(tmp);
        ofs << input;
    }

    SHealth shealth;
    shealth.CalculateBmi(tmp);

    std::ostringstream oss;
    oss.setf(std::ios::fixed, std::ios::floatfield);
    oss << std::setprecision(1);

    oss << "BMI Ratios for 20s (%)\n";
    oss << "Underweight: " << shealth.GetBmiRatio(20, 100) << "\n";
    oss << "Normal: "      << shealth.GetBmiRatio(20, 200) << "\n";
    oss << "Overweight: "  << shealth.GetBmiRatio(20, 300) << "\n";
    oss << "Obesity: "     << shealth.GetBmiRatio(20, 400) << "\n";

    Approvals::verify(oss.str());
    std::remove(tmp);
}

TEST(SHealthApprovalTests, EmptyAgeGroupApproval)
{
    auto approvalsSubdir = Approvals::useApprovalsSubdirectory("approvals");

    // 특정 연령대(40대)에 데이터가 없는 경우 테스트
    std::string input =
        "id,age,weight,height\n"
        "1,25,70,170\n"  // 20대
        "2,35,65,175\n"  // 30대
        "3,55,80,180\n"; // 50대 (40대 데이터 없음)

    const char* tmp = "temp_empty_age_group.csv";
    {
        std::ofstream ofs(tmp);
        ofs << input;
    }

    SHealth shealth;
    shealth.CalculateBmi(tmp);

    std::ostringstream oss;
    oss.setf(std::ios::fixed, std::ios::floatfield);
    oss << std::setprecision(1);

    oss << "BMI Ratios for 40s (%) - Should be all zeros\n";
    oss << "Underweight: " << shealth.GetBmiRatio(40, 100) << "\n";
    oss << "Normal: "      << shealth.GetBmiRatio(40, 200) << "\n";
    oss << "Overweight: "  << shealth.GetBmiRatio(40, 300) << "\n";
    oss << "Obesity: "     << shealth.GetBmiRatio(40, 400) << "\n";

    Approvals::verify(oss.str());
    std::remove(tmp);
}

TEST(SHealthApprovalTests, AllZeroWeightsApproval)
{
    auto approvalsSubdir = Approvals::useApprovalsSubdirectory("approvals");

    // 모든 체중이 0인 경우 (보간 불가능)
    std::string input =
        "id,age,weight,height\n"
        "1,25,0,170\n"
        "2,25,0,175\n"
        "3,25,0,180\n";

    const char* tmp = "temp_all_zero_weights.csv";
    {
        std::ofstream ofs(tmp);
        ofs << input;
    }

    SHealth shealth;
    shealth.CalculateBmi(tmp);

    std::ostringstream oss;
    oss.setf(std::ios::fixed, std::ios::floatfield);
    oss << std::setprecision(1);

    oss << "BMI Ratios for 20s (%) - All zero weights case\n";
    oss << "Underweight: " << shealth.GetBmiRatio(20, 100) << "\n";
    oss << "Normal: "      << shealth.GetBmiRatio(20, 200) << "\n";
    oss << "Overweight: "  << shealth.GetBmiRatio(20, 300) << "\n";
    oss << "Obesity: "     << shealth.GetBmiRatio(20, 400) << "\n";

    Approvals::verify(oss.str());
    std::remove(tmp);
}

TEST(SHealthApprovalTests, InvalidDataHandlingApproval)
{
    auto approvalsSubdir = Approvals::useApprovalsSubdirectory("approvals");

    // 유효하지 않은 데이터가 포함된 경우
    std::string input =
        "id,age,weight,height\n"
        "1,25,70,170\n"      // 정상 데이터
        "2,-5,65,175\n"      // 음수 나이 (무시됨)
        "3,200,80,180\n"     // 비현실적 나이 (무시됨)
        "4,30,65,0\n"        // 키 0 (무시됨)
        "5,35,1500,175\n"    // 비현실적 체중 (0으로 처리 후 보간)
        "6,35,75,175\n";     // 정상 데이터

    const char* tmp = "temp_invalid_data.csv";
    {
        std::ofstream ofs(tmp);
        ofs << input;
    }

    SHealth shealth;
    int count = shealth.CalculateBmi(tmp);

    std::ostringstream oss;
    oss.setf(std::ios::fixed, std::ios::floatfield);
    oss << std::setprecision(1);

    oss << "Valid records processed: " << count << "\n";
    oss << "BMI Ratios for 20s (%)\n";
    oss << "Underweight: " << shealth.GetBmiRatio(20, 100) << "\n";
    oss << "Normal: "      << shealth.GetBmiRatio(20, 200) << "\n";
    oss << "Overweight: "  << shealth.GetBmiRatio(20, 300) << "\n";
    oss << "Obesity: "     << shealth.GetBmiRatio(20, 400) << "\n";

    oss << "BMI Ratios for 30s (%)\n";
    oss << "Underweight: " << shealth.GetBmiRatio(30, 100) << "\n";
    oss << "Normal: "      << shealth.GetBmiRatio(30, 200) << "\n";
    oss << "Overweight: "  << shealth.GetBmiRatio(30, 300) << "\n";
    oss << "Obesity: "     << shealth.GetBmiRatio(30, 400) << "\n";

    Approvals::verify(oss.str());
    std::remove(tmp);
}

TEST(SHealthApprovalTests, ExactBoundaryValuesApproval)
{
    auto approvalsSubdir = Approvals::useApprovalsSubdirectory("approvals");

    // BMI 경계값들의 정확한 분류 테스트
    std::string input =
        "id,age,weight,height\n"
        "1,25,18.5,100\n"   // BMI = 18.5 (저체중 경계)
        "2,25,18.6,100\n"   // BMI = 18.6 (정상체중 하한)
        "3,25,22.9,100\n"   // BMI = 22.9 (정상체중 상한)
        "4,25,23.0,100\n"   // BMI = 23.0 (과체중 하한)
        "5,25,24.9,100\n"   // BMI = 24.9 (과체중 상한)
        "6,25,25.0,100\n"   // BMI = 25.0 (어떤 카테고리에도 포함되지 않음)
        "7,25,25.1,100\n";  // BMI = 25.1 (비만 하한)

    const char* tmp = "temp_exact_boundaries.csv";
    {
        std::ofstream ofs(tmp);
        ofs << input;
    }

    SHealth shealth;
    int count = shealth.CalculateBmi(tmp);

    std::ostringstream oss;
    oss.setf(std::ios::fixed, std::ios::floatfield);
    oss << std::setprecision(1);

    oss << "Total records: " << count << "\n";
    oss << "BMI Ratios for 20s (%)\n";
    oss << "Underweight: " << shealth.GetBmiRatio(20, 100) << "\n";
    oss << "Normal: "      << shealth.GetBmiRatio(20, 200) << "\n";
    oss << "Overweight: "  << shealth.GetBmiRatio(20, 300) << "\n";
    oss << "Obesity: "     << shealth.GetBmiRatio(20, 400) << "\n";

    Approvals::verify(oss.str());
    std::remove(tmp);
}

