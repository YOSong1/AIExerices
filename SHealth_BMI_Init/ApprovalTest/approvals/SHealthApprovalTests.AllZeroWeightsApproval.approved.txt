BMI Ratios for 20s (%) - All zero weights case
Underweight: 100.0
Normal: 0.0
Overweight: 0.0
Obesity: 0.0

